rustworkx-core/src/bipartite_coloring.rs

```diff
@@ -0,0 +1,1122 @@
+        g.remove_matching(&matching);
+        odd_degree_matching = Some(matching);
+    }
+
+    // Split graph into two regular bipartite multigraphs of half-degree
+    let (h1, mut h2) = rbmg_split_into_two(&g);
+
+    // Recursively color h1
+    let h1_coloring = rbmg_edge_color(&h1);
+
+    // Transfer some matchings from H1 to H2 to make the degree of H2 to be a power of 2
+    let r = h2.degree.next_power_of_two();
+    let num_classes_to_move = r - h2.degree;
+
+    for matching in h1_coloring.iter().take(num_classes_to_move) {
+        h2.add_matching(matching);
+    }
+
+    // Edge-color h2 by recursive splitting
+    let h2_coloring = rbmg_edge_color_when_power_of_two(&h2);
+
+    // The matching consists of h2 colors, remaining h1 colors and the removed
+    // matching (if had odd degree)
+    coloring = h2_coloring;
+    for matching in h1_coloring.iter().skip(num_classes_to_move) {
+        coloring.push(matching.clone());
+    }
+    if let Some(matching) = odd_degree_matching {
+        coloring.push(matching);
+    }
+
+    if coloring.len() != g0.degree {
+        panic!("Wrong output coloring length");
+    }
+    coloring
+}
+
+/// Color edges of a bipartite graph.
+///
+/// The implementation is based on the following paper:
+///
+/// Noga Alon. "A simple algorithm for edge-coloring bipartite multigraphs".
+/// Inf. Process. Lett. 85(6), (2003).
+/// <https://www.tau.ac.il/~nogaa/PDFS/lex2.pdf>
+///
+/// The input to the algorithm is a bipartite graph and an explicit partition of
+/// its nodes into "left" and "right" nodes. The output is an assignment of colors
+/// to the edges so that no two incident edges have the same color. The algorithm
+/// runs in time `O (m log m)`, where `m` is the number of edges of the graph.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on. The graph is
+///     assumed to be bipartite.
+/// * `l_nodes` - The vector containing the "left" nodes of the graph.
+/// * `r_nodes` - The vector containing the "right" nodes of the graph.
+///
+/// # Example
+/// ```rust
+/// use petgraph::graph::Graph;
+/// use petgraph::graph::{EdgeIndex, NodeIndex};
+/// use petgraph::Undirected;
+/// use rustworkx_core::dictmap::*;
+/// use rustworkx_core::bipartite_coloring::bipartite_edge_color_given_partition;
+///
+/// let edge_list = vec![(0, 1), (1, 2), (2, 3), (3, 0)];
+/// let graph = Graph::<(), (), Undirected>::from_edges(&edge_list);
+/// let l_nodes = vec![NodeIndex::new(0), NodeIndex::new(2)];
+/// let r_nodes = vec![NodeIndex::new(1), NodeIndex::new(3)];
+/// let colors = bipartite_edge_color_given_partition(&graph, &l_nodes, &r_nodes);
+/// let expected_colors: DictMap<EdgeIndex, usize> = [
+///     (EdgeIndex::new(0), 0),
+///     (EdgeIndex::new(1), 1),
+///     (EdgeIndex::new(2), 0),
+///     (EdgeIndex::new(3), 1),
+/// ]
+/// .into_iter()
+/// .collect();
+/// assert_eq!(colors, expected_colors);
+/// ```
+pub fn bipartite_edge_color_given_partition<G>(
```

Comment:
I don't think that having such a function is really useful (I did think so at the beginning but have changed my mind since then). Technically, we could avoid calling `two_color` if a partition is already given, but `two_color` is super-fast (since it's essentially DFS). Moreover, exposing this function to the users would probably mean making sure to handle the invalid input (i.e. making sure that the given partition is indeed a bipartite partition). 

The ability to preset some colors (as in #1060) sounds very useful and it may be a good idea to similarly extend `greedy_edge_color`. Unfortunately I don't think we can have something similar for bipartite or for Misra-Gries edge-coloring algorithms.

Reply:
The reason I asked is because this is a public function so we're exposing it to end users of the rustworkx-core library. I guess the follow on question is should this be `pub` or not?